src/twister2/generate_tests_plugin.py

```diff
@@ -0,0 +1,163 @@
+    if 'specification' not in metafunc.fixturenames:
+        return
+
+    twister_config = metafunc.config.twister_config
+
+    platforms_list: list[PlatformSpecification] = [
+        platform for platform in twister_config.platforms
+        if platform.identifier in twister_config.default_platforms
+    ]
+    spec_file_path: Path = Path(metafunc.definition.fspath.dirname) / TEST_SPEC_FILE_NAME
+    scenarios = get_scenarios_from_fixture(metafunc)
+    assert spec_file_path.exists(), f'There is no specification file for the test: {spec_file_path}'
+    if not scenarios:
+        scenarios = get_scenarios_from_yaml(spec_file_path)
+    variants = itertools.product(platforms_list, scenarios)
+    params: list[pytest.param] = []
+    for variant in variants:
+        v = Variant(*variant)
+        params.append(
+            pytest.param(v, marks=pytest.mark.platform(v.platform.identifier), id=str(v))
+        )
+
+    # using indirect=True to inject value from `specification` fixture instead of param
+    metafunc.parametrize(
+        'specification', params, scope='function', indirect=True
+    )
+
+
+@pytest.fixture(scope='function')
+def specification() -> None:
+    """Injects a test specification from yaml file to a test item"""
+    # The body of this function is empty because it is only used to
+    # inform pytest that we want to generate parametrized tests for
+    # a test function which uses this fixture
+
+
+def pytest_configure(config: pytest.Config):
+    config.addinivalue_line(
+        'markers', 'build_specification(names="scenario1,scenario2"): select scenarios to build'
+    )
+
+
+def generate_yaml_test_specification_for_item(item: pytest.Item, variant: Variant) -> YamlTestSpecification:
+    """Add test specification from yaml file to test item."""
+    logger.debug('Adding test specification to item "%s"', item.nodeid)
+    scenario: str = variant.scenario
+    platform: PlatformSpecification = variant.platform
+
+    spec_path: Path = item.path.parent.joinpath(TEST_SPEC_FILE_NAME)
+    assert spec_path.exists(), f'Spec file does not exist: {spec_path}'
+    test_directory_path: Path = item.path.parent
+    rootpath: Path = item.config.rootpath
+
+    raw_spec: dict = safe_load_yaml(spec_path)
+    tests: dict = extract_tests(raw_spec)
+
+    try:
+        test_spec_dict = tests[scenario]
+    except KeyError:
+        msg = f'There is no specification for {scenario} in file {spec_path}'
+        logger.error(msg)
+        raise TwisterConfigurationException(msg)
+
+    test_spec_dict['name'] = item.name
+    test_spec_dict['original_name'] = item.originalname
+    test_spec_dict['source_dir'] = test_directory_path
+    test_spec_dict['platform'] = platform.identifier
+    test_spec_dict['build_name'] = scenario
+    test_spec_dict['rel_to_base_path'] = Path.relative_to(test_spec_dict['source_dir'], rootpath)
+
+    test_spec = YamlTestSpecification(**test_spec_dict)
+
+    # add markers to the test
```

Comment:
Could be this adding markers unified somehow (at least partially) with these ones:
https://github.com/zephyrproject-rtos/twister/blob/3e63f4703ebab2630fe89e5df4767cc415da4472/src/twister2/yaml_test_function.py#L31-L38
And by "unified" I mean creating one function/method to apply those markers (`tags`, `slow`, `skip`, etc.). I guess that it is not trivial, but on the other hand now, we have two places with similar code and during modification one of then it is necessary to remember to modify second one.

Reply:
I have created single function which adds markers to to pytest.Funtion and pytest.Item. Now we have only one place where this code exists.